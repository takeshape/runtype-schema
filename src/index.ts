export * from './jsonSchemaToRuntype';
export * from './runtypeToJsonSchema';
