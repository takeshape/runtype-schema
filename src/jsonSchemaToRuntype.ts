import {JSONSchema6Definition, JSONSchema7Definition} from 'json-schema';
import {Runtype} from 'runtypes';
import {create} from 'runtypes/lib/runtype';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';

export function createAjv(): Ajv {
  const ajv = new Ajv();
  addFormats(ajv);
  return ajv;
}

const validator = createAjv();

export const jsonSchemaToRuntype = <T>(schema: JSONSchema6Definition | JSONSchema7Definition): Runtype<T> => {
  const validate = validator.compile(schema);
  // @ts-expect-error
  return create((v: any) => {
    if (!validate(v)) {
      throw Error(`v is not a valid ${JSON.stringify(schema)}`);
    }
    return v;
  }, {});
};
