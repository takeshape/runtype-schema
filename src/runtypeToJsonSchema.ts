import {Runtype, Reflect} from 'runtypes';
import {JSONSchema7Definition} from 'json-schema';

const isJsonScalar = (v: any) =>
  ['number', 'string', 'boolean', 'object'].includes(typeof v) &&
  (typeof v !== 'number' || (!Number.isNaN(v) && Number.isFinite(v))) &&
  (typeof v !== 'object' || v === null);

export const runtypeToJsonSchema = <T>(t: Runtype<T> | Reflect): JSONSchema7Definition => {
  const r = t.reflect;
  switch (r.tag) {
    case 'symbol':
    case 'function':
    case 'constraint':
    case 'instanceof':
    case 'brand':
      throw Error(`${r.tag} can't be converted to json schema`);

    case 'always':
      return true;
    case 'never':
      return false;
    case 'void':
      return {enum: [null]};

    case 'boolean':
    case 'number':
    case 'string':
      return {type: r.tag};

    case 'literal':
      if (!isJsonScalar(r.value)) throw Error(`can't make json schema for literal "${JSON.stringify(r.value)}"`);
      if (r.value === null) return {type: 'null'};
      return {enum: [r.value as any]};

    case 'array':
      return {
        type: 'array',
        items: runtypeToJsonSchema(r.element)
      };

    case 'record':
      return {
        type: 'object',
        required: Object.keys(r.fields),
        properties: Object.keys(r.fields).reduce<Record<string, JSONSchema7Definition>>(
          (props, f) => ({...props, [f]: runtypeToJsonSchema(r.fields[f])}),
          {}
        )
      };

    case 'partial':
      return {
        anyOf: [
          {type: 'string'},
          {type: 'number'},
          {type: 'boolean'},
          {type: 'array'},
          {
            type: 'object',
            properties: Object.keys(r.fields).reduce<Record<string, JSONSchema7Definition>>(
              (props, f) => ({
                ...props,
                [f]: runtypeToJsonSchema(r.fields[f])
              }),
              {}
            )
          }
        ]
      };

    case 'dictionary':
      return {
        type: 'object',
        additionalProperties: runtypeToJsonSchema(r.value)
      };

    case 'tuple':
      return {
        type: 'array',
        minItems: r.components.length,
        maxItems: r.components.length,
        items: r.components.map(runtypeToJsonSchema)
      };

    case 'union':
      return {
        anyOf: r.alternatives.map(runtypeToJsonSchema)
      };

    case 'intersect':
      return {
        allOf: r.intersectees.map(runtypeToJsonSchema)
      };

    default:
      // istanbul ignore next
      throw Error(`invalid runtype ${(r as any).tag}`);
  }
};
