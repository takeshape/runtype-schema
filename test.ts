import * as runtype from 'runtypes';
import fc from 'fast-check';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';

export function createAjv(): Ajv {
  const ajv = new Ajv();
  addFormats(ajv);
  return ajv;
}

const validator = createAjv();

import {runtypeToJsonSchema} from './src/runtypeToJsonSchema';
import {jsonSchemaToRuntype} from './src/jsonSchemaToRuntype';

const validJson = fc.jsonObject;
const jsonScalar = () =>
  fc
    // @ts-expect-error
    .oneof<string | number | boolean | null>(fc.constant(null), fc.string(), fc.double(), fc.nat(), fc.boolean())
    .filter(v => typeof v !== 'number' || (!Number.isNaN(v) && Number.isFinite(v)));

const testdata: Array<{
  name: string;
  runtype?: any;
  runtypeGenerator?: any;
}> = [
  {
    name: 'string',
    runtype: runtype.String
  },
  {name: 'number', runtype: runtype.Number},
  {name: 'boolean', runtype: runtype.Boolean},
  {name: 'always', runtype: runtype.Always},
  {name: 'never', runtype: runtype.Never},
  {name: 'void', runtype: runtype.Void},
  {
    name: 'literal',
    runtype: runtype.Literal(null),
    runtypeGenerator: jsonScalar().map(s => runtype.Literal(s as any))
  },
  {
    name: 'array',
    runtype: runtype.Array(runtype.Literal(null)),
    // @ts-expect-error
    runtypeGenerator: jsonScalar().map(v => runtype.Array(runtype.Literal(v)))
  },
  {
    name: 'record',
    runtype: runtype.Record({
      // @ts-expect-error
      a: runtype.Literal('a'),
      // @ts-expect-error
      b: runtype.Literal('b')
    }),
    runtypeGenerator: fc.tuple(jsonScalar(), jsonScalar()).map(([a, b]) =>
      runtype.Record({
        // @ts-expect-error
        a: runtype.Literal(a),
        // @ts-expect-error
        b: runtype.Literal(b)
      })
    )
  },
  {
    name: 'partial',
    runtype: runtype.Partial({
      // @ts-expect-error
      a: runtype.Literal('a'),
      // @ts-expect-error
      b: runtype.Literal('b')
    }),
    runtypeGenerator: fc
      .tuple(jsonScalar(), jsonScalar())
      // @ts-expect-error
      .map(([a, b]) => runtype.Partial({a: runtype.Literal(a), b: runtype.Literal(b)}))
  },
  {
    name: 'dictionary',
    runtype: runtype.Dictionary(runtype.Number),
    // @ts-expect-error
    runtypeGenerator: jsonScalar().map(v => runtype.Dictionary(runtype.Literal(v)))
  },
  {
    name: 'tuple',
    runtype: runtype.Tuple(runtype.Literal('a'), runtype.Literal('b'))
  },
  {
    name: 'union',
    runtype: runtype.Union(runtype.Number, runtype.String)
  },
  {
    name: 'intersect',
    runtype: runtype.Intersect(runtype.String, runtype.Number)
  }
];

const invalidTestdata: {
  name: string;
  runtype: any;
}[] = [
  {name: 'symbol', runtype: runtype.Symbol},
  {name: 'function', runtype: runtype.Function},
  {
    name: 'constraint',
    runtype: runtype.Constraint(runtype.String, s => s.length > 5)
  },
  {name: 'instance of', runtype: runtype.InstanceOf(Object)},
  {name: 'brand', runtype: runtype.Brand('brand', runtype.String)},
  {name: 'invalid literal', runtype: runtype.Literal([] as any)}
];

describe('runtype to json schema', () => {
  describe('invalid types', () => {
    invalidTestdata.forEach(({name, runtype: r}) => {
      test(name, () => {
        expect(() => runtypeToJsonSchema(r)).toThrow();
      });
    });
  });

  testdata.forEach(({name, runtype: r, runtypeGenerator}) => {
    describe(name, () => {
      if (r) {
        test('snapshot', () => {
          const schema = runtypeToJsonSchema(r);
          expect(schema).toMatchSnapshot(name);
        });
      }

      if (r || runtypeGenerator) {
        describe('compatibility', () => {
          if (r) {
            test('given runtype gives same result on any value', () => {
              const schema = runtypeToJsonSchema(r);
              const r2 = jsonSchemaToRuntype(schema);
              try {
                const validate = validator.compile(schema);
                fc.assert(
                  fc.property(validJson(), v => {
                    expect(validate(v)).toBe(r.guard(v));
                    expect(r2.guard(v)).toBe(r.guard(v));
                  })
                );
              } catch (e) {
                console.log(JSON.stringify(schema, null, 2));
                throw e;
              }
            });
          }

          if (runtypeGenerator) {
            test('generated runtype gives same result on any value', () => {
              fc.assert(
                fc.property(runtypeGenerator, fc.jsonObject(), (r: runtype.Runtype, v) => {
                  const schema = runtypeToJsonSchema(r);
                  const r2 = jsonSchemaToRuntype(schema);
                  const validate = validator.compile(schema);
                  expect(validate(v)).toBe(r.guard(v));
                  expect(r2.guard(v)).toBe(r.guard(v));
                })
              );
            });
          }
        });
      }
    });
  });
});
