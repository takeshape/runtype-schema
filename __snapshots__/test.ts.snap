// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtype to json schema always snapshot: always 1`] = `true`;

exports[`runtype to json schema array snapshot: array 1`] = `
Object {
  "items": Object {
    "type": "null",
  },
  "type": "array",
}
`;

exports[`runtype to json schema boolean snapshot: boolean 1`] = `
Object {
  "type": "boolean",
}
`;

exports[`runtype to json schema dictionary snapshot: dictionary 1`] = `
Object {
  "additionalProperties": Object {
    "type": "number",
  },
  "type": "object",
}
`;

exports[`runtype to json schema intersect snapshot: intersect 1`] = `
Object {
  "allOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
  ],
}
`;

exports[`runtype to json schema literal snapshot: literal 1`] = `
Object {
  "type": "null",
}
`;

exports[`runtype to json schema never snapshot: never 1`] = `false`;

exports[`runtype to json schema number snapshot: number 1`] = `
Object {
  "type": "number",
}
`;

exports[`runtype to json schema partial snapshot: partial 1`] = `
Object {
  "anyOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
    Object {
      "type": "array",
    },
    Object {
      "properties": Object {
        "a": Object {
          "enum": Array [
            "a",
          ],
        },
        "b": Object {
          "enum": Array [
            "b",
          ],
        },
      },
      "type": "object",
    },
  ],
}
`;

exports[`runtype to json schema record snapshot: record 1`] = `
Object {
  "properties": Object {
    "a": Object {
      "enum": Array [
        "a",
      ],
    },
    "b": Object {
      "enum": Array [
        "b",
      ],
    },
  },
  "required": Array [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`runtype to json schema string snapshot: string 1`] = `
Object {
  "type": "string",
}
`;

exports[`runtype to json schema tuple snapshot: tuple 1`] = `
Object {
  "items": Array [
    Object {
      "enum": Array [
        "a",
      ],
    },
    Object {
      "enum": Array [
        "b",
      ],
    },
  ],
  "maxItems": 2,
  "minItems": 2,
  "type": "array",
}
`;

exports[`runtype to json schema union snapshot: union 1`] = `
Object {
  "anyOf": Array [
    Object {
      "type": "number",
    },
    Object {
      "type": "string",
    },
  ],
}
`;

exports[`runtype to json schema void snapshot: void 1`] = `
Object {
  "enum": Array [
    null,
  ],
}
`;
